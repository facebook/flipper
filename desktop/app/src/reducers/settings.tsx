/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @format
 */

import {Actions} from './index';
import os from 'os';
import electron from 'electron';

export enum Tristate {
  True,
  False,
  Unset,
}

export type Settings = {
  androidHome: string;
  enableAndroid: boolean;
  enableIOS: boolean;
  enablePhysicalIOS: boolean;
  /**
   * If unset, this will assume the value of the GK setting.
   * Note that this setting has no effect in the open source version
   * of Flipper.
   */
  enablePrefetching: Tristate;
  idbPath: string;
  jsApps: {
    webAppLauncher: {
      url: string;
      height: number;
      width: number;
    };
  };
  reactNative: {
    shortcuts: {
      enabled: boolean;
      reload: string;
      openDevMenu: string;
    };
  };
  darkMode: string;
  showWelcomeAtStartup: boolean;
  suppressPluginErrors: boolean;
};

export type Action =
  | {type: 'INIT'}
  | {
      type: 'UPDATE_SETTINGS';
      payload: Settings;
    };

export const DEFAULT_ANDROID_SDK_PATH = getDefaultAndroidSdkPath();

const initialState: Settings = {
  androidHome: getDefaultAndroidSdkPath(),
  enableAndroid: true,
  enableIOS: os.platform() === 'darwin',
  enablePhysicalIOS: os.platform() === 'darwin',
  enablePrefetching: Tristate.Unset,
  idbPath: '/usr/local/bin/idb',
  jsApps: {
    webAppLauncher: {
      url: 'http://localhost:8888',
      height: 600,
      width: 800,
    },
  },
  reactNative: {
    shortcuts: {
      enabled: false,
      reload: 'Alt+Shift+R',
      openDevMenu: 'Alt+Shift+D',
    },
  },
  darkMode: 'auto',
  showWelcomeAtStartup: true,
  suppressPluginErrors: false,
};

export default function reducer(
  state: Settings = initialState,
  action: Actions,
): Settings {
  if (action.type === 'UPDATE_SETTINGS') {
    return action.payload;
  }
  return state;
}

export function updateSettings(settings: Settings): Action {
  return {
    type: 'UPDATE_SETTINGS',
    payload: settings,
  };
}

function getDefaultAndroidSdkPath() {
  return os.platform() === 'win32' ? getWindowsSdkPath() : '/opt/android_sdk';
}

function getWindowsSdkPath() {
  const app = electron.app || electron.remote.app;
  return `${app.getPath('home')}\\AppData\\Local\\android\\sdk`;
}
