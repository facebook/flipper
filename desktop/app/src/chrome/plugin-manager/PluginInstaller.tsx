/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @format
 */

import {Layout, theme} from 'flipper-plugin';
import {LoadingIndicator, TableRows, ManagedTable, Glyph} from '../../ui';
import React, {useCallback, useState, useEffect} from 'react';
import {reportPlatformFailures, reportUsage} from '../../utils/metrics';
import reloadFlipper from '../../utils/reloadFlipper';
import {registerInstalledPlugins} from '../../reducers/plugins';
import {
  UpdateResult,
  getInstalledPlugins,
  getUpdatablePlugins,
  removePlugin,
  UpdatablePluginDetails,
  InstalledPluginDetails,
} from 'flipper-plugin-lib';
import {installPluginFromNpm} from 'flipper-plugin-lib';
import {State as AppState} from '../../reducers';
import {connect} from 'react-redux';
import {Dispatch, Action} from 'redux';
import PluginPackageInstaller from './PluginPackageInstaller';
import {Toolbar} from 'flipper-plugin';
import {Alert, Button, Input, Tooltip, Typography} from 'antd';

const {Text, Link} = Typography;

const TAG = 'PluginInstaller';

const columnSizes = {
  name: '25%',
  version: '10%',
  description: 'flex',
  install: '15%',
};

const columns = {
  name: {
    value: 'Name',
  },
  version: {
    value: 'Version',
  },
  description: {
    value: 'Description',
  },
  install: {
    value: '',
  },
};

type PropsFromState = {
  installedPlugins: Map<string, InstalledPluginDetails>;
};

type DispatchFromProps = {
  refreshInstalledPlugins: () => void;
};

type OwnProps = {
  autoHeight: boolean;
};

type Props = OwnProps & PropsFromState & DispatchFromProps;

const defaultProps: OwnProps = {
  autoHeight: false,
};

const PluginInstaller = function ({
  refreshInstalledPlugins,
  installedPlugins,
  autoHeight,
}: Props) {
  const [restartRequired, setRestartRequired] = useState(false);
  const [query, setQuery] = useState('');

  const onInstall = useCallback(async () => {
    refreshInstalledPlugins();
    setRestartRequired(true);
  }, [refreshInstalledPlugins]);

  const rows = useNPMSearch(query, onInstall, installedPlugins);
  const restartApp = useCallback(() => {
    reloadFlipper();
  }, []);

  return (
    <Layout.Container gap height={500}>
      {restartRequired && (
        <Alert
          onClick={restartApp}
          type="error"
          message="To apply the changes, Flipper needs to reload. Click here to reload!"
          style={{cursor: 'pointer'}}
        />
      )}
      <Toolbar>
        <Input.Search
          onChange={(e) => setQuery(e.target.value)}
          value={query}
          placeholder="Search Flipper plugins..."
        />
      </Toolbar>
      <ManagedTable
        rowLineHeight={28}
        floating={false}
        multiline
        columnSizes={columnSizes}
        columns={columns}
        highlightableRows={false}
        highlightedRows={new Set()}
        autoHeight={autoHeight}
        rows={rows}
      />
      <PluginPackageInstaller onInstall={onInstall} />
    </Layout.Container>
  );
};

function InstallButton(props: {
  name: string;
  version: string;
  onInstall: () => void;
  updateStatus: UpdateResult;
}) {
  type InstallAction =
    | {kind: 'Install'; error?: string}
    | {kind: 'Waiting'}
    | {kind: 'Remove'; error?: string}
    | {kind: 'Update'; error?: string};

  const catchError =
    (actionKind: 'Install' | 'Remove' | 'Update', fn: () => Promise<void>) =>
    async () => {
      try {
        await fn();
      } catch (err) {
        console.error(err);
        setAction({kind: actionKind, error: err.toString()});
      }
    };

  const mkInstallCallback = (action: 'Install' | 'Update') =>
    catchError(action, async () => {
      reportUsage(
        action === 'Install' ? `${TAG}:install` : `${TAG}:update`,
        undefined,
        props.name,
      );
      setAction({kind: 'Waiting'});

      await installPluginFromNpm(props.name);

      props.onInstall();
      setAction({kind: 'Remove'});
    });

  const performInstall = useCallback(mkInstallCallback('Install'), [
    props.name,
    props.version,
  ]);

  const performUpdate = useCallback(mkInstallCallback('Update'), [
    props.name,
    props.version,
  ]);

  const performRemove = useCallback(
    catchError('Remove', async () => {
      reportUsage(`${TAG}:remove`, undefined, props.name);
      setAction({kind: 'Waiting'});
      await removePlugin(props.name);
      props.onInstall();
      setAction({kind: 'Install'});
    }),
    [props.name],
  );

  const [action, setAction] = useState<InstallAction>(
    props.updateStatus.kind === 'update-available'
      ? {kind: 'Update'}
      : props.updateStatus.kind === 'not-installed'
      ? {kind: 'Install'}
      : {kind: 'Remove'},
  );

  if (action.kind === 'Waiting') {
    return <LoadingIndicator size={16} />;
  }
  if ((action.kind === 'Install' || action.kind === 'Remove') && action.error) {
  }
  const button = (
    <Button
      size="small"
      type={action.kind !== 'Remove' ? 'primary' : undefined}
      onClick={() => {
        switch (action.kind) {
          case 'Install':
            reportPlatformFailures(performInstall(), `${TAG}:install`);
            break;
          case 'Remove':
            reportPlatformFailures(performRemove(), `${TAG}:remove`);
            break;
          case 'Update':
            reportPlatformFailures(performUpdate(), `${TAG}:update`);
            break;
        }
      }}>
      {action.kind}
    </Button>
  );

  if (action.error) {
    const glyph = (
      <Glyph color={theme.warningColor} size={16} name="caution-triangle" />
    );
    return (
      <Layout.Horizontal gap>
        <Tooltip
          placement="leftBottom"
          title={`Something went wrong: ${action.error}`}
          children={glyph}
        />
        {button}
      </Layout.Horizontal>
    );
  } else {
    return button;
  }
}

function useNPMSearch(
  query: string,
  onInstall: () => void,
  installedPlugins: Map<string, InstalledPluginDetails>,
): TableRows {
  useEffect(() => {
    reportUsage(`${TAG}:open`);
  }, []);

  const [searchResults, setSearchResults] = useState<UpdatablePluginDetails[]>(
    [],
  );

  const createRow = useCallback(
    (h: UpdatablePluginDetails) => ({
      key: h.name,
      columns: {
        name: {
          value: <Text ellipsis>{h.name.replace(/^flipper-plugin-/, '')}</Text>,
        },
        version: {
          value: <Text ellipsis>{h.version}</Text>,
          align: 'flex-end' as 'flex-end',
        },
        description: {
          value: (
            <Layout.Horizontal center gap>
              <Text ellipsis>{h.description}</Text>
              <Link href={`https://yarnpkg.com/en/package/${h.name}`}>
                <Glyph
                  color={theme.textColorActive}
                  name="info-circle"
                  size={16}
                />
              </Link>
            </Layout.Horizontal>
          ),
        },
        install: {
          value: (
            <InstallButton
              name={h.name}
              version={h.version}
              onInstall={onInstall}
              updateStatus={h.updateStatus}
            />
          ),
          align: 'center' as 'center',
        },
      },
    }),
    [onInstall],
  );

  useEffect(() => {
    (async () => {
      let cancelled = false;
      const updatablePlugins = await reportPlatformFailures(
        getUpdatablePlugins(query),
        `${TAG}:queryIndex`,
      );
      if (cancelled) {
        return;
      }
      setSearchResults(updatablePlugins);
      // Clean up: if query changes while we're searching, abandon results.
      return () => {
        cancelled = true;
      };
    })();
  }, [query, installedPlugins]);

  const rows = searchResults.map(createRow);
  return rows;
}

PluginInstaller.defaultProps = defaultProps;

export default connect<PropsFromState, DispatchFromProps, OwnProps, AppState>(
  ({plugins: {installedPlugins}}) => ({
    installedPlugins,
  }),
  (dispatch: Dispatch<Action<any>>) => ({
    refreshInstalledPlugins: () => {
      getInstalledPlugins().then((plugins) =>
        dispatch(registerInstalledPlugins(plugins)),
      );
    },
  }),
)(PluginInstaller);
