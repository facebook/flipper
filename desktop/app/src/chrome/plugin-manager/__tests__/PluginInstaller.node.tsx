/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @format
 */

jest.mock('flipper-plugin-lib');

import {default as PluginInstaller} from '../PluginInstaller';
import React from 'react';
import {render, waitFor} from '@testing-library/react';
import configureStore from 'redux-mock-store';
import {Provider} from 'react-redux';
import type {PluginDetails} from 'flipper-plugin-lib';
import {getUpdatablePlugins, UpdatablePluginDetails} from 'flipper-plugin-lib';
import {Store} from '../../../reducers';
import {mocked} from 'ts-jest/utils';

const getUpdatablePluginsMock = mocked(getUpdatablePlugins);

function getStore(installedPlugins: PluginDetails[] = []): Store {
  return configureStore([])({
    application: {sessionId: 'mysession'},
    plugins: {installedPlugins},
  }) as Store;
}

const samplePluginDetails1: UpdatablePluginDetails = {
  name: 'flipper-plugin-hello',
  entry: './test/index.js',
  version: '0.1.0',
  specVersion: 2,
  pluginType: 'client',
  main: 'dist/bundle.js',
  dir: '/Users/mock/.flipper/thirdparty/flipper-plugin-sample1',
  source: 'src/index.js',
  id: 'Hello',
  title: 'Hello',
  description: 'World?',
  isBundled: false,
  isActivatable: true,
  updateStatus: {
    kind: 'not-installed',
    version: '0.1.0',
  },
};

const samplePluginDetails2: UpdatablePluginDetails = {
  name: 'flipper-plugin-world',
  entry: './test/index.js',
  version: '0.2.0',
  specVersion: 2,
  pluginType: 'client',
  main: 'dist/bundle.js',
  dir: '/Users/mock/.flipper/thirdparty/flipper-plugin-sample2',
  source: 'src/index.js',
  id: 'World',
  title: 'World',
  description: 'Hello?',
  isBundled: false,
  isActivatable: true,
  updateStatus: {
    kind: 'not-installed',
    version: '0.2.0',
  },
};

const SEARCH_RESULTS = [samplePluginDetails1, samplePluginDetails2];

afterEach(() => {
  getUpdatablePluginsMock.mockClear();
});

test('load PluginInstaller list', async () => {
  getUpdatablePluginsMock.mockReturnValue(Promise.resolve(SEARCH_RESULTS));
  const component = (
    <Provider store={getStore()}>
      <PluginInstaller
        // Bit ugly to have this as an effectively test-only option, but
        // without, we rely on height information from Electron which we don't
        // have, causing no items to be rendered.
        autoHeight
      />
    </Provider>
  );
  const {container, getByText} = render(component);
  await waitFor(() => getByText('hello'));
  expect(getUpdatablePluginsMock.mock.calls.length).toBe(1);
  expect(container).toMatchSnapshot();
});

test('load PluginInstaller list with one plugin installed', async () => {
  getUpdatablePluginsMock.mockReturnValue(
    Promise.resolve([
      {...samplePluginDetails1, updateStatus: {kind: 'up-to-date'}},
      samplePluginDetails2,
    ]),
  );
  const store = getStore([samplePluginDetails1]);
  const component = (
    <Provider store={store}>
      <PluginInstaller
        // Bit ugly to have this as an effectively test-only option, but
        // without, we rely on height information from Electron which we don't
        // have, causing no items to be rendered.
        autoHeight
      />
    </Provider>
  );
  const {container, getByText} = render(component);
  await waitFor(() => getByText('hello'));
  expect(getUpdatablePluginsMock.mock.calls.length).toBe(1);
  expect(container).toMatchSnapshot();
});
