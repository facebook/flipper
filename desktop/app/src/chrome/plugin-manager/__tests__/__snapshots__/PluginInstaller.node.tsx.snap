// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`load PluginInstaller list 1`] = `
<div>
  <div
    class="css-1v0y38i-Container e1hsqii15"
    height="500"
  >
    <div
      class="css-1lxv8hi-Container-Horizontal-SandyToolbarContainer e1ecpah20"
    >
      <span
        class="ant-input-group-wrapper ant-input-search"
      >
        <span
          class="ant-input-wrapper ant-input-group"
        >
          <input
            class="ant-input"
            placeholder="Search Flipper plugins..."
            type="text"
            value=""
          />
          <span
            class="ant-input-group-addon"
          >
            <button
              class="ant-btn ant-btn-icon-only ant-input-search-button"
              type="button"
            >
              <span
                aria-label="search"
                class="anticon anticon-search"
                role="img"
              >
                <svg
                  aria-hidden="true"
                  data-icon="search"
                  fill="currentColor"
                  focusable="false"
                  height="1em"
                  viewBox="64 64 896 896"
                  width="1em"
                >
                  <path
                    d="M909.6 854.5L649.9 594.8C690.2 542.7 712 479 712 412c0-80.2-31.3-155.4-87.9-212.1-56.6-56.7-132-87.9-212.1-87.9s-155.5 31.3-212.1 87.9C143.2 256.5 112 331.8 112 412c0 80.1 31.3 155.5 87.9 212.1C256.5 680.8 331.8 712 412 712c67 0 130.6-21.8 182.7-62l259.7 259.6a8.2 8.2 0 0011.6 0l43.6-43.5a8.2 8.2 0 000-11.6zM570.4 570.4C528 612.7 471.8 636 412 636s-116-23.3-158.4-65.6C211.3 528 188 471.8 188 412s23.3-116.1 65.6-158.4C296 211.3 352.2 188 412 188s116.1 23.2 158.4 65.6S636 352.2 636 412s-23.3 116.1-65.6 158.4z"
                  />
                </svg>
              </span>
            </button>
          </span>
        </span>
      </span>
    </div>
    <div
      class="css-p5h61d-View-FlexBox-FlexColumn-Container esta8x30"
      tabindex="0"
    >
      <div
        class="css-18abd42-View-FlexBox-FlexColumn ecr18to0"
      >
        <div
          class="css-1hdusim-View-FlexBox-FlexRow-TableHeadContainer ejga3101"
        >
          <div
            class="css-it9ar6-TableHeadColumnContainer ejga3100"
            title="name"
            width="25%"
          >
            <div
              class="ejga3103 css-x4q70f-InteractiveContainer-TableHeaderColumnInteractive e14xwmxq0"
              style="z-index: auto; right: 0px; bottom: 0px; width: 100%; height: 100%;"
            >
              <div
                class="css-1054ash-TableHeaderColumnContainer ejga3102"
              >
                Name
              </div>
            </div>
          </div>
          <div
            class="css-1m9d124-TableHeadColumnContainer ejga3100"
            title="version"
            width="10%"
          >
            <div
              class="ejga3103 css-x4q70f-InteractiveContainer-TableHeaderColumnInteractive e14xwmxq0"
              style="z-index: auto; right: 0px; bottom: 0px; width: 100%; height: 100%;"
            >
              <div
                class="css-1054ash-TableHeaderColumnContainer ejga3102"
              >
                Version
              </div>
            </div>
          </div>
          <div
            class="css-qmwopt-TableHeadColumnContainer ejga3100"
            title="description"
            width="flex"
          >
            <div
              class="ejga3103 css-x4q70f-InteractiveContainer-TableHeaderColumnInteractive e14xwmxq0"
              style="z-index: auto; right: 0px; bottom: 0px; width: 100%; height: 100%;"
            >
              <div
                class="css-1054ash-TableHeaderColumnContainer ejga3102"
              >
                Description
              </div>
            </div>
          </div>
          <div
            class="css-ux5muk-TableHeadColumnContainer ejga3100"
            title="install"
            width="15%"
          >
            <div
              class="ejga3103 css-x4q70f-InteractiveContainer-TableHeaderColumnInteractive e14xwmxq0"
              style="z-index: auto; right: 0px; bottom: 0px; width: 100%; height: 100%;"
            >
              <div
                class="css-1054ash-TableHeaderColumnContainer ejga3102"
              >
                
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="css-p5h61d-View-FlexBox-FlexColumn-Container esta8x30"
      >
        <div
          class="css-18abd42-View-FlexBox-FlexColumn ecr18to0"
        >
          <div
            class="css-hg3ptm-View-FlexBox-FlexRow-TableBodyRowContainer ehuguum1"
            data-key="flipper-plugin-hello"
          >
            <div
              class="css-hdg6vt-TableBodyColumnContainer ehuguum0"
              title=""
              width="25%"
            >
              <span
                class="ant-typography ant-typography-ellipsis ant-typography-single-line ant-typography-ellipsis-single-line"
              >
                hello
              </span>
            </div>
            <div
              class="css-1rdsust-TableBodyColumnContainer ehuguum0"
              title=""
              width="10%"
            >
              <span
                class="ant-typography ant-typography-ellipsis ant-typography-single-line ant-typography-ellipsis-single-line"
              >
                0.1.0
              </span>
            </div>
            <div
              class="css-1bw2bjn-TableBodyColumnContainer ehuguum0"
              title=""
              width="flex"
            >
              <div
                class="css-s1wsbn-Container-Horizontal e1hsqii14"
              >
                <span
                  class="ant-typography ant-typography-ellipsis ant-typography-single-line ant-typography-ellipsis-single-line"
                >
                  World?
                </span>
                <a
                  class="ant-typography"
                  href="https://yarnpkg.com/en/package/flipper-plugin-hello"
                >
                  <div
                    class="css-1kmzf9v-ColoredIconCustom e528dze0"
                    color="var(--light-color-button-active)"
                    size="16"
                    src="https://facebook.com/assets/?name=info-circle&variant=filled&size=16&set=facebook_icons&density=1x"
                  />
                </a>
              </div>
            </div>
            <div
              class="css-4dx1ef-TableBodyColumnContainer ehuguum0"
              title=""
              width="15%"
            >
              <button
                class="ant-btn ant-btn-primary ant-btn-sm"
                type="button"
              >
                <span>
                  Install
                </span>
              </button>
            </div>
          </div>
          <div
            class="css-hg3ptm-View-FlexBox-FlexRow-TableBodyRowContainer ehuguum1"
            data-key="flipper-plugin-world"
          >
            <div
              class="css-hdg6vt-TableBodyColumnContainer ehuguum0"
              title=""
              width="25%"
            >
              <span
                class="ant-typography ant-typography-ellipsis ant-typography-single-line ant-typography-ellipsis-single-line"
              >
                world
              </span>
            </div>
            <div
              class="css-1rdsust-TableBodyColumnContainer ehuguum0"
              title=""
              width="10%"
            >
              <span
                class="ant-typography ant-typography-ellipsis ant-typography-single-line ant-typography-ellipsis-single-line"
              >
                0.2.0
              </span>
            </div>
            <div
              class="css-1bw2bjn-TableBodyColumnContainer ehuguum0"
              title=""
              width="flex"
            >
              <div
                class="css-s1wsbn-Container-Horizontal e1hsqii14"
              >
                <span
                  class="ant-typography ant-typography-ellipsis ant-typography-single-line ant-typography-ellipsis-single-line"
                >
                  Hello?
                </span>
                <a
                  class="ant-typography"
                  href="https://yarnpkg.com/en/package/flipper-plugin-world"
                >
                  <div
                    class="css-1kmzf9v-ColoredIconCustom e528dze0"
                    color="var(--light-color-button-active)"
                    size="16"
                    src="https://facebook.com/assets/?name=info-circle&variant=filled&size=16&set=facebook_icons&density=1x"
                  />
                </a>
              </div>
            </div>
            <div
              class="css-4dx1ef-TableBodyColumnContainer ehuguum0"
              title=""
              width="15%"
            >
              <button
                class="ant-btn ant-btn-primary ant-btn-sm"
                type="button"
              >
                <span>
                  Install
                </span>
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="css-1lxv8hi-Container-Horizontal-SandyToolbarContainer e1ecpah20"
    >
      <div
        class="css-1spj5hr-View-FlexBox-FlexRow-Container ersmi542"
      >
        <input
          class="css-sli06x-Input-FileInputBox ersmi540"
          placeholder="Specify path to a Flipper package or just drag and drop it here..."
          type="text"
          value=""
        />
        <div
          class="css-ccdckn-View-FlexBox-FlexRow-GlyphContainer ersmi541"
        >
          <img
            alt="dots-3-circle"
            class="ersmi543 css-6iptsk-ColoredIconBlack-CenteredGlyph e528dze1"
            size="16"
            src="https://facebook.com/assets/?name=dots-3-circle&variant=outline&size=16&set=facebook_icons&density=1x"
            title="Open file selection dialog"
          />
        </div>
        <div
          class="css-ccdckn-View-FlexBox-FlexRow-GlyphContainer ersmi541"
        >
          <div
            class="css-auhar3-TooltipContainer e1abcqbd0"
          >
            <div
              class="ersmi543 css-1qsl9s4-ColoredIconCustom-CenteredGlyph e528dze0"
              color="#D79651"
              size="16"
              src="https://facebook.com/assets/?name=caution-triangle&variant=filled&size=16&set=facebook_icons&density=1x"
            />
          </div>
        </div>
      </div>
      <div
        class="css-5ukfaz-View-FlexBox-FlexRow-ButtonContainer evd5j491"
      >
        <div
          class="css-wospjg-View-FlexBox-FlexRow epz0qe20"
        >
          <button
            class="ant-btn ant-btn-primary"
            disabled=""
            type="button"
          >
            <span>
              Install
            </span>
          </button>
          <div
            class="css-170i4ha-View-FlexBox-FlexRow-ErrorGlyphContainer evd5j490"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`load PluginInstaller list with one plugin installed 1`] = `
<div>
  <div
    class="css-1v0y38i-Container e1hsqii15"
    height="500"
  >
    <div
      class="css-1lxv8hi-Container-Horizontal-SandyToolbarContainer e1ecpah20"
    >
      <span
        class="ant-input-group-wrapper ant-input-search"
      >
        <span
          class="ant-input-wrapper ant-input-group"
        >
          <input
            class="ant-input"
            placeholder="Search Flipper plugins..."
            type="text"
            value=""
          />
          <span
            class="ant-input-group-addon"
          >
            <button
              class="ant-btn ant-btn-icon-only ant-input-search-button"
              type="button"
            >
              <span
                aria-label="search"
                class="anticon anticon-search"
                role="img"
              >
                <svg
                  aria-hidden="true"
                  data-icon="search"
                  fill="currentColor"
                  focusable="false"
                  height="1em"
                  viewBox="64 64 896 896"
                  width="1em"
                >
                  <path
                    d="M909.6 854.5L649.9 594.8C690.2 542.7 712 479 712 412c0-80.2-31.3-155.4-87.9-212.1-56.6-56.7-132-87.9-212.1-87.9s-155.5 31.3-212.1 87.9C143.2 256.5 112 331.8 112 412c0 80.1 31.3 155.5 87.9 212.1C256.5 680.8 331.8 712 412 712c67 0 130.6-21.8 182.7-62l259.7 259.6a8.2 8.2 0 0011.6 0l43.6-43.5a8.2 8.2 0 000-11.6zM570.4 570.4C528 612.7 471.8 636 412 636s-116-23.3-158.4-65.6C211.3 528 188 471.8 188 412s23.3-116.1 65.6-158.4C296 211.3 352.2 188 412 188s116.1 23.2 158.4 65.6S636 352.2 636 412s-23.3 116.1-65.6 158.4z"
                  />
                </svg>
              </span>
            </button>
          </span>
        </span>
      </span>
    </div>
    <div
      class="css-p5h61d-View-FlexBox-FlexColumn-Container esta8x30"
      tabindex="0"
    >
      <div
        class="css-18abd42-View-FlexBox-FlexColumn ecr18to0"
      >
        <div
          class="css-1hdusim-View-FlexBox-FlexRow-TableHeadContainer ejga3101"
        >
          <div
            class="css-it9ar6-TableHeadColumnContainer ejga3100"
            title="name"
            width="25%"
          >
            <div
              class="ejga3103 css-x4q70f-InteractiveContainer-TableHeaderColumnInteractive e14xwmxq0"
              style="z-index: auto; right: 0px; bottom: 0px; width: 100%; height: 100%;"
            >
              <div
                class="css-1054ash-TableHeaderColumnContainer ejga3102"
              >
                Name
              </div>
            </div>
          </div>
          <div
            class="css-1m9d124-TableHeadColumnContainer ejga3100"
            title="version"
            width="10%"
          >
            <div
              class="ejga3103 css-x4q70f-InteractiveContainer-TableHeaderColumnInteractive e14xwmxq0"
              style="z-index: auto; right: 0px; bottom: 0px; width: 100%; height: 100%;"
            >
              <div
                class="css-1054ash-TableHeaderColumnContainer ejga3102"
              >
                Version
              </div>
            </div>
          </div>
          <div
            class="css-qmwopt-TableHeadColumnContainer ejga3100"
            title="description"
            width="flex"
          >
            <div
              class="ejga3103 css-x4q70f-InteractiveContainer-TableHeaderColumnInteractive e14xwmxq0"
              style="z-index: auto; right: 0px; bottom: 0px; width: 100%; height: 100%;"
            >
              <div
                class="css-1054ash-TableHeaderColumnContainer ejga3102"
              >
                Description
              </div>
            </div>
          </div>
          <div
            class="css-ux5muk-TableHeadColumnContainer ejga3100"
            title="install"
            width="15%"
          >
            <div
              class="ejga3103 css-x4q70f-InteractiveContainer-TableHeaderColumnInteractive e14xwmxq0"
              style="z-index: auto; right: 0px; bottom: 0px; width: 100%; height: 100%;"
            >
              <div
                class="css-1054ash-TableHeaderColumnContainer ejga3102"
              >
                
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="css-p5h61d-View-FlexBox-FlexColumn-Container esta8x30"
      >
        <div
          class="css-18abd42-View-FlexBox-FlexColumn ecr18to0"
        >
          <div
            class="css-hg3ptm-View-FlexBox-FlexRow-TableBodyRowContainer ehuguum1"
            data-key="flipper-plugin-hello"
          >
            <div
              class="css-hdg6vt-TableBodyColumnContainer ehuguum0"
              title=""
              width="25%"
            >
              <span
                class="ant-typography ant-typography-ellipsis ant-typography-single-line ant-typography-ellipsis-single-line"
              >
                hello
              </span>
            </div>
            <div
              class="css-1rdsust-TableBodyColumnContainer ehuguum0"
              title=""
              width="10%"
            >
              <span
                class="ant-typography ant-typography-ellipsis ant-typography-single-line ant-typography-ellipsis-single-line"
              >
                0.1.0
              </span>
            </div>
            <div
              class="css-1bw2bjn-TableBodyColumnContainer ehuguum0"
              title=""
              width="flex"
            >
              <div
                class="css-s1wsbn-Container-Horizontal e1hsqii14"
              >
                <span
                  class="ant-typography ant-typography-ellipsis ant-typography-single-line ant-typography-ellipsis-single-line"
                >
                  World?
                </span>
                <a
                  class="ant-typography"
                  href="https://yarnpkg.com/en/package/flipper-plugin-hello"
                >
                  <div
                    class="css-1kmzf9v-ColoredIconCustom e528dze0"
                    color="var(--light-color-button-active)"
                    size="16"
                    src="https://facebook.com/assets/?name=info-circle&variant=filled&size=16&set=facebook_icons&density=1x"
                  />
                </a>
              </div>
            </div>
            <div
              class="css-4dx1ef-TableBodyColumnContainer ehuguum0"
              title=""
              width="15%"
            >
              <button
                class="ant-btn ant-btn-sm"
                type="button"
              >
                <span>
                  Remove
                </span>
              </button>
            </div>
          </div>
          <div
            class="css-hg3ptm-View-FlexBox-FlexRow-TableBodyRowContainer ehuguum1"
            data-key="flipper-plugin-world"
          >
            <div
              class="css-hdg6vt-TableBodyColumnContainer ehuguum0"
              title=""
              width="25%"
            >
              <span
                class="ant-typography ant-typography-ellipsis ant-typography-single-line ant-typography-ellipsis-single-line"
              >
                world
              </span>
            </div>
            <div
              class="css-1rdsust-TableBodyColumnContainer ehuguum0"
              title=""
              width="10%"
            >
              <span
                class="ant-typography ant-typography-ellipsis ant-typography-single-line ant-typography-ellipsis-single-line"
              >
                0.2.0
              </span>
            </div>
            <div
              class="css-1bw2bjn-TableBodyColumnContainer ehuguum0"
              title=""
              width="flex"
            >
              <div
                class="css-s1wsbn-Container-Horizontal e1hsqii14"
              >
                <span
                  class="ant-typography ant-typography-ellipsis ant-typography-single-line ant-typography-ellipsis-single-line"
                >
                  Hello?
                </span>
                <a
                  class="ant-typography"
                  href="https://yarnpkg.com/en/package/flipper-plugin-world"
                >
                  <div
                    class="css-1kmzf9v-ColoredIconCustom e528dze0"
                    color="var(--light-color-button-active)"
                    size="16"
                    src="https://facebook.com/assets/?name=info-circle&variant=filled&size=16&set=facebook_icons&density=1x"
                  />
                </a>
              </div>
            </div>
            <div
              class="css-4dx1ef-TableBodyColumnContainer ehuguum0"
              title=""
              width="15%"
            >
              <button
                class="ant-btn ant-btn-primary ant-btn-sm"
                type="button"
              >
                <span>
                  Install
                </span>
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="css-1lxv8hi-Container-Horizontal-SandyToolbarContainer e1ecpah20"
    >
      <div
        class="css-1spj5hr-View-FlexBox-FlexRow-Container ersmi542"
      >
        <input
          class="css-sli06x-Input-FileInputBox ersmi540"
          placeholder="Specify path to a Flipper package or just drag and drop it here..."
          type="text"
          value=""
        />
        <div
          class="css-ccdckn-View-FlexBox-FlexRow-GlyphContainer ersmi541"
        >
          <img
            alt="dots-3-circle"
            class="ersmi543 css-6iptsk-ColoredIconBlack-CenteredGlyph e528dze1"
            size="16"
            src="https://facebook.com/assets/?name=dots-3-circle&variant=outline&size=16&set=facebook_icons&density=1x"
            title="Open file selection dialog"
          />
        </div>
        <div
          class="css-ccdckn-View-FlexBox-FlexRow-GlyphContainer ersmi541"
        >
          <div
            class="css-auhar3-TooltipContainer e1abcqbd0"
          >
            <div
              class="ersmi543 css-1qsl9s4-ColoredIconCustom-CenteredGlyph e528dze0"
              color="#D79651"
              size="16"
              src="https://facebook.com/assets/?name=caution-triangle&variant=filled&size=16&set=facebook_icons&density=1x"
            />
          </div>
        </div>
      </div>
      <div
        class="css-5ukfaz-View-FlexBox-FlexRow-ButtonContainer evd5j491"
      >
        <div
          class="css-wospjg-View-FlexBox-FlexRow epz0qe20"
        >
          <button
            class="ant-btn ant-btn-primary"
            disabled=""
            type="button"
          >
            <span>
              Install
            </span>
          </button>
          <div
            class="css-170i4ha-View-FlexBox-FlexRow-ErrorGlyphContainer evd5j490"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;
